File: python/n-queens.py
# 51. N 皇后 - Hard
# https://leetcode-cn.com/problems/n-queens/

from typing import List

def solveQueens(n: int) -> List[List[str]]:
    ans = []
    rowCol = [[False] * n] * n
    forward = [False] * n
    backward = [False] * n
    queens = [-1] * n
    def isValid(row: int, col: int) -> bool:
        if rowCol[row][col] or forward[row + col] or backward[abs(col - row)]:
            return False
        return True
    def backtrack(row: int):
        if row == n:
            board = list()
            for i in range(n):
                row[queens[i]] = "Q"
                board.append("".join(row))
                row[queens[i]] = "."
            ans.append(board)
            return
        for col in range(n):
            if not isValid(row, col):
                continue
            rowCol[row][col] = True
            forward[col + row] = True
            backward[abs(row-col)] = True
            queens[row] = col
            backtrack(row + 1)
            rowCol[row][col] = False
            forward[col + row] = False
            backward[abs(row-col)] = False
    
    row = ["." * n] * n 
    backtrack(0)
    return ans

    

File: java/n-queens.java
// 
// 
package java;

import java.util.ArrayList;
import java.util.List;

class NQueens {

    List<List<String>> ans;
    boolean[][] rowCol;
    boolean[] forward;
    boolean[] backward;
    
    public List<List<String>> solveQueens(int n) {
        rowCol = new boolean[n][n];
        forward = new boolean[2 * n];
        backward = new boolean[n];
        ans = new ArrayList<>(n);
        String[][] board = new String[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                board[i][j] = ".";
            }
        }
        backtrack(board, 0, n);
        return ans;
    }

    public void backtrack(String[][] board, int row, int n) {
        if (row == n) {
            List<String> list = new ArrayList<>();
            for (String[] line : board) {
                list.add(String.join("", line));
            }
            ans.add(list);
        }
        for (int col = 0; col < n; col++) {
            if (!isValid(row, col)) continue;
            rowCol[row][col] = true;
            forward[row + col] = true;
            backward[Math.abs(col - row)] = true;
            board[row][col] = "Q";
            backtrack(board, row + 1, n);
            rowCol[row][col] = false;
            forward[row + col] = false;
            backward[Math.abs(col - row)] = false;
            board[row][col] = ".";
        }
    }

    private boolean isValid(int row, int col) {
        int v = Math.abs(col - row);
        return !(rowCol[row][col] || forward[row + col] || backward[v]);
    }
}
